### PR TITLE
Closes #677 fix non-direct descendant ordering in selector

### LINKED ISSUE #677: Non-direct descendant ordering in selector
**Describe the bug**
At the moment, doing something like `div button` doesn't work, as we only expect direct descendants. Also in reality, if you do `div > button`, `div span button` will also match, which is wrong.

**To Reproduce**
1. Create an action, add a `div button` selector
 

**Expected behavior**
Should match events

**Screenshots**
image
image

**Desktop (optional: please complete the following information):**
 - OS:
 - Browser: 
 - Version: 

**Smartphone (optional: please complete the following information):**
 - Device: 
 - OS: 
 - Browser 
 - Version 

**Additional context**

### PR DESCRIPTION
**Pull request by timgl**
## Changes

- Make sure direct descendants (`div > button`) doesn't match incorrectly (like `div span button`)
- Allow non-direct descendant selectors (`div button` to match `div span button`)
- Refactor the parsing of selectors into something slightly more sturdy.

## Checklist
- [x] All querysets/queries filter by Team (if applicable)
- [x] Backend tests (if applicable)

### CODE REVIEW

**Review by mariusandra — CHANGES_REQUESTED** [review]
Hi, I found a bug. 

It's always a bad idea to modify an array/list that you're actively iterating over: `del tags[index - 1]`. This causes a bug now as well and skips an element. I added a test case and `selector2` crashes.

I recommend also creating an exhaustive test suite for the `Selector` and `SelectorPart` code, not just for `Event`s.

I didn't test the `_element_subquery` code yet.

**Comment by timgl** [comment]
Good catch! We didn't even need that del anymore so that's good.

Agree, have added a test suite. It was already implicitly tested by `test_filter_with_selectors` but good to test separately.

File: posthog/models/event.py [hunk 1]
```diff
@@ -20,67 +20,96 @@
...
+    def __init__(self, selector: str):
+        self.parts = []
+        tags = re.split(" ", selector)
+        tags.reverse()
+        for index, tag in enumerate(tags):
+            if tag == ">":
+                continue
+            direct_descendant = False
+            if tags[index - 1] == ">":
+                direct_descendant = True
```
Comment (mariusandra) [inline posthog/models/event.py:69]:
```
tags = ['div', '>', 'span', '>', 'a']
tags.reverse()
index = 0
tags[index - 1] == 'div'
```

Is that the right thing to check? It feels like it's checking the wrong thing and there might be some room for abuse here, though I couldn't come up with something that would make this crash... other than a `"> div span"` or  `"div span >"` type incorrect query potentially having the `direct_descendant` flag wrong.

Still might be worth pushing in an `if index > 0 and tags[...` clause?

**Review by mariusandra — APPROVED** [review]
I found one more thing, which *might* cause an issue... and/or chould be flagged with a "# yeah but it works" comment. The rest seems fine.

File: posthog/models/event.py (hunk 1 above)
Comment (timgl) [reply posthog/models/event.py]: Added

**Pull request merged**